src/SAML2/XML/samlp/RequestedAuthnContext.php

```diff
@@ -0,0 +1,142 @@
+        $authnContextClassRef = Utils::xpQuery($xml, './saml_assertion:AuthnContextClassRef');
+
+        /** @var \DOMElement[] $authnContextDeclRef */
+        $authnContextDeclRef = Utils::xpQuery($xml, './saml_assertion:AuthnContextDeclRef');
+
+        Assert::oneOf(
+            [],
+            [$authnContextClassRef, $authnContextDeclRef],
+            'You need either AuthnContextClassRef or AuthnContextDeclRef, not both'
+        );
+
+        foreach ($authnContextClassRef as $classRef) {
+            $this->requestedAuthnContexts[] = new AuthnContextClassRef($classRef);
+        }
+
+        foreach ($authnContextDeclRef as $declRef) {
+            $this->requestedAuthnContexts[] = new AuthnContextDeclRef($declRef);
+        }
+
+        if ($xml->hasAttribute('Comparison')) {
+            $this->Comparison = $xml->getAttribute('Comparison');
+        }
+    }
+
+
+    /**
+     * Collect the value of the requestedAuthnContexts-property
+     *
+     * @return (\SAML2\XML\saml\AuthnContextClassRef|\SAML2\XML\saml\AuthnContextDeclRef)[]
+     */
+    public function getRequestedAuthnContexts(): array
+    {
+        return $this->requestedAuthnContexts;
+    }
+
+
+    /**
+     * Set the value of the requestedAuthnContexts-property
+     *
+     * @param (\SAML2\XML\saml\AuthnContextClassRef|\SAML2\XML\saml\AuthnContextDeclRef)[] $requestedAuthnContexts
+     * @return void
+     */
+    public function setRequestedAuthnContexts(array $requestedAuthnContexts): void
+    {
+        $this->requestedAuthnContexts = $requestedAuthnContexts;
+    }
+
+
+    /**
+     * Collect the value of the Comparison-property
+     *
+     * @return string|null
+     */
+    public function getComparison(): ?string
+    {
+        return $this->Comparison;
+    }
+
+
+    /**
+     * Set the value of the Comparison-property
+     *
+     * @param string|null $comparison
+     * @return void
+     */
+    public function setComparison(?string $comparison): void
+    {
+        $this->Comparison = $comparison;
+    }
+
+
+    /**
+     * Convert this RequestedAuthnContext to XML.
+     *
+     * @param \DOMElement $parent The element we should append this RequestedAuthnContext to.
+     * @return \DOMElement
+     */
+    public function toXML(DOMElement $parent): DOMElement
+    {
+        $e = $parent->ownerDocument->createElementNS(Constants::NS_SAMLP, 'samlp:RequestedAuthnContext');
+        $parent->appendChild($e);
+
+        $types = array_unique(array_map('gettype', $this->requestedAuthnContexts));
+        Assert::maxCount(
+            $types,
+            1,
+            'You need either AuthnContextClassRef or AuthnContextDeclRef, not both'
+        );
+
+        foreach ($this->requestedAuthnContexts as $context) {
+            $e->appendChild($e->ownerDocument->importNode($context->toXML(), true));
+        }
+
+        if (isset($this->Comparison)) {
+            Assert::oneOf($this->Comparison, ['exact', 'minimum', 'maximum', 'better']);
```

Comment:
This assertion is probably better in `setComparison()`. That way, if it fails, it fails early and in the place where you did something wrong (assigning an erroneous value), rather than when converting to XML.

Reply:
If we do that, we have to use the setter-method in the constructor as well.. That way we can never import XML with an unspecified value

Reply:
So to add to this, we basically need to decide whether we do our sanity/compliancy checks on setting the values or when generating the output XML..  Previously we tested the output to make sure it's compliant, but personally I'm in favour of catching issues on the input-side..

Reply:
Agreed, I think the sooner, the better. I also thought about calling the setter in the constructor. If _unspecified_ is legal (which I don't know), then the constructor could only call the setter if there was a value specified.